third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -118,6 +133,21 @@ std::vector<ComputationClient::DataPtr> PjRtComputationClient::TransferToServer(
   return datas;
 }
 
+ComputationClient::DataPtr PjRtComputationClient::TransferShardsToServer(
+    absl::Span<const TensorSource> tensor_shards, std::string device,
```

Comment:
Do you need the `device` and `shape` arguments here? They're contained in `TensorSource` for each tensor: https://github.com/pytorch/xla/blob/6e42e7cb3af01d9f8909e112a3be0148a87acad0/third_party/xla_client/computation_client.h#L129-L130

Reply:
Yes, because they are the original (unpartitioned) shape and device. Each tensor source is a shard and their source and device are specific to the shard.